### DISCUSSION TITLE
FirmwareUpdater → Socket-CAN Not responding

### DISCUSSION THREAD

**Discussion by RGalviati42** [robotology/community#623] [Q&A] [answered]
Dear all,

I am experiencing some troubles trying to update the firmware of the MTB4 boards. In detail I installed on my machine (Ubuntu 22.04) the Firmware Updater tool and all the required drivers for CAN communication (CAN interface : CAN-USB by ESD). As a check I used CANutils to communicate with the board via terminal and it sends and receives correctly. Nonetheless, whenever I try to run the FirmwareUpdater and to discover the nodes of the network I get the following ERROR (See figures) message telling me the Hardware is either busy or not connected.

```console
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
"CFW_CAN" - "0"
"CFW_CAN" - "1"
"ESD_CAN" - "5"
"ESD_CAN" - "6"
"SOCKETCAN" - "can0"
"ETH" - "10.0.1.104:3333"
IP address FOUND in .ini file, Using : "10.0.1.104"
Port Number FOUND in .ini file, Using : "3333"
ACE_SOCK_Dgram: Cannot assign requested address
ACE_SOCK_Dgram_Bcast::mk_broadcast: setsockopt failed: Operation not supported
ACE_SOCK_Dgram_Bcast: Operation not supported
[ERROR] |yarp.os.YarpPluginSettings| Cannot find "socketcan" plugin (not built in, and no .ini file found for it)Check that YARP_DATA_DIRS leads to at least one directory with plugins/socketcan.ini or share/yarp/plugins/socketcan.ini in it
[ERROR] |yarp.dev.PolyDriver|socketcan| Could not find device <socketcan>
FirmwareUpdaterCore::getCanBoardsFromDriver(): Init driver failed - Hardware busy or not connected?!
[ERROR] |yarp.os.YarpPluginSettings| Cannot find "socketcan" plugin (not built in, and no .ini file found for it)Check that YARP_DATA_DIRS leads to at least one directory with plugins/socketcan.ini or share/yarp/plugins/socketcan.ini in it
[ERROR] |yarp.dev.PolyDriver|socketcan| Could not find device <socketcan>
FirmwareUpdaterCore::getCanBoardsFromDriver(): Init driver failed - Hardware busy or not connected?!
```

If anyone can help I'll be grateful.
Thanks in advance.
Riccardo.

![MicrosoftTeams-image (20)](https://user-images.githubusercontent.com/132926449/236870788-840d37c5-e237-4966-9b83-5a0f2c1292f8.png)

**Comment by Nicogene** [comment, marked as answer]
Hi @RGalviati42!

Your log is telling you that is missing the `icub-main` device for communicating via can, [`socketcan`](https://github.com/robotology/icub-main/tree/master/src/libraries/icubmod/socketCan).

For enabling it you need to configure icub-main with `-DENABLE_icubmod_socketcan:BOOL=ON`.

cc @davidetome @marcoaccame @sgiraz

**Reply by davidetome** [reply]
> Hi @RGalviati42!
> 
> Your log is telling you that is missing the `icub-main` device for communicating via can, [`socketcan`](https://github.com/robotology/icub-main/tree/master/src/libraries/icubmod/socketCan).
> 
> For enabling it you need to configure icub-main with `-DENABLE_icubmod_socketcan:BOOL=ON`.
> 
> cc @davidetome @marcoaccame @sgiraz

or if you're using robotology-superbuild simply turn ON ROBOTOLOGY_ENABLE_ICUB_HEAD
